pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+   @overload
+   def foo(x: str) -> str: ...
+
+Here is how type checkers should handle usage of this library:
+
+.. code-block:: python
+
+   from library import Ham  # error: Use of deprecated class Ham. Use Spam instead.
+
+   import library
+
+   library.norwegian_gray(1)  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.
+   map(library.norwegian_gray, [1, 2, 3])  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.
+
+   library.foo(1)  # error: Use of deprecated overload for foo. Only str will be allowed.
+   library.foo("x")  # no error
+
+Runtime behavior
+----------------
+
+At runtime, the decorator sets an attribute ``.__deprecated__`` on the decorated
+object. The value of the attribute is the message passed to the decorator.
+The decorator returns the original object. Notably, it does not issue a runtime
+``DeprecationWarning``.
+
+Type checker behavior
+---------------------
+
+This PEP does not specify exactly how type checkers should present deprecation
+diagnostics to their users. However, some users (e.g., application developers
+targeting only a specific version of Python) may not care about deprecations,
+while others (e.g., library developers who want their library to remain
+compatible with future versions of Python) would want to catch any use of
+deprecated functionality in their CI pipeline. Therefore, it is recommended
+that type checker provide configuration options that cover both use cases.
+
+Backwards compatibility
+=======================
+
+Creating a new decorator poses no backwards compatibility concerns.
+As with all new typing functionality, the ``@deprecated`` decorator
+will be added to the ``typing_extensions`` module, enabling its use
+in older versions of Python.
+
+How to teach this
+=================
+
+For users who encounter deprecation warnings in their IDE or type
+checker output, the messages they receive should be clear and self-explanatory.
+Usage of the ``@deprecated`` decorator will be an advanced feature
+mostly relevant to library authors. The decorator should be mentioned
+in relevant documentation (e.g., :pep:`387` and the :exc:`DeprecationWarning`
+documentation) as an additional way to mark deprecated functionality.
+
+Reference Implementation
+========================
+
+TODO
+
+Rejected ideas
+==============
+
+Runtime warnings
+----------------
+
+Users might expect usage of the ``@deprecated`` decorator to issue a
+``DeprecationWarning`` at runtime. However, this would raise a number of
+thorny issues:
+
+* When the decorator is applied to a class or an overload, the warning
+  would not be raised as expected.
+* Users may want to control the ``warn`` call in more detail (e.g.,
+  changing the warning class).
+* ``typing.py`` generally aims to avoid affecting runtime behavior.
+
+Users who want to use ``@deprecated`` while also issuing a runtime warning
+can use the ``if TYPE_CHECKING:`` idiom, for example:
```

Comment:
I hope this doesn't become too popular, or we'd be better off doing this ourselves (it would be tricky to add later -- if people add other mechanisms they would get double warnings). And we _do_ have `reveal_type()`, which prints to stderr at runtime.

Reply:
@hauntsaninja also brought this up above. I'm now thinking of adding a `runtime=True` kwarg to the decorator that makes it also throw a runtime warning.

Reply:
On the one hand, users will probably expect some runtime behaviour, and will probably be surprised if there isn't a runtime effect.

On the other hand, adding runtime behaviour feels like we might be stepping on the toes of third-party packages that have spent a lot of time thinking about this kind of thing, like https://github.com/tantale/deprecated.

(I don't really have a solution here or a strong preference, just noting that it might be an issue.)

Reply:
Yeah, for what it worth, it does `raise` some potential considerations, but if people are otherwise going to be doing ugly, boilerplately and verbose things, it's better to just provide the functionality built in now, IMO.

Reply:
Pyright also really doesn't like code in the `else` branch of `if TYPE_CHECKING`: https://github.com/microsoft/pyright/discussions/2330

Reply:
Will bring this up as a point of discussion when starting threads.